### ISSUE TITLE
Question: what parameters are being adjusted for speed/accuracy tradeoff in benchmarks?

### ISSUE THREAD

**Issue by Michiel29** [facebookresearch/faiss#1653]
I am attempting to get good performance for an index in range of 100m - 1 billion vectors. As a guideline, I'm using the benchmark at https://github.com/facebookresearch/faiss/wiki/Indexing-1G-vectors. Using the high-performing index strings on my own data, I am nowhere near reproducing the listed speeds. 

For example, using faiss-cpu from conda and the index string "OPQ32_64,IVF65536_HNSW32,PQ32" on a 50m x 256 dataset with 32 threads on a 72 core machine, I get approximately 10^3 QPS (other index strings from benchmarks give similar results). 

One possibility is that I am all the way at the accuracy part of the accuracy-speed tradeoff. However, neither the benchmark documentation nor bench_all_ivf.py seem to show what parameters are actually being changed. As far as I'm aware the only parameter for IVF indices is nprobe (I have nprobe = 5), which I imagine shouldn't result in a 10^2-10^3 order speed difference?

Thank you for your time.

**Comment by mdouze** [comment]
The parameters are tuned automatically, see https://github.com/facebookresearch/faiss/wiki/Index-IO,-cloning-and-hyper-parameter-tuning#auto-tuning-the-runtime-parameters

`bench_all_ivf.py` *does* display the actual parameters, for example for the index type you mention you can see the hyperparameters with

```
>>> import faiss
>>> index = faiss.index_factory(256, "OPQ32_64,IVF65536_HNSW32,PQ32")
>>> params = faiss.ParameterSpace()
>>> params.initialize(index)
>>> params.display()
ParameterSpace, 3 parameters, 6760 combinations:
   nprobe: [ 1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048, 4096]
   quantizer_efSearch: [ 4, 8, 16, 32, 64, 128, 256, 512]
   ht: [ 2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42, 44, 46, 48, 50, 52, 54, 56, 58, 60, 62, 64, 66, 68, 70, 72, 74, 76, 78, 80, 82, 84, 86, 88, 90, 92, 94, 96, 98, 100, 102, 104, 106, 108, 110, 112, 114, 116, 118, 120, 122, 124, 126, 128, 256]
```

**Comment by mdouze** [comment]
Here is an example of bench_all_ivf search: https://gist.github.com/mdouze/105a0b933c9455697c8d6f16e82449ba
As reported, the times to perform a query are between 0.00656 ms and 2.53783, so there *is* a 300x factor between slowest and fastest parameters (not all parameters are all useful in practice).

**Comment by Michiel29** [comment]
Ah, they are contained in the ParameterSpace. That makes sense, thank you for elaborating. Reproducing those operating points on my data, I get wildly different results - 1 ms/q is about the fastest I get, and 4 ms/q is about the slowest I get, so only a factor 4 difference (and this is for a smaller dataset, so I would expect 0.02 ms/q or faster). 

I will try to see what I get on the actual bench_all_ivf, but is it expected that similarly sized data should get such extremely different results? A factor 100 seems like a lot, and I am not sure what could affect performance given similar:

- dataset size
- cores & num_threads
- index string
- run-time parameters

**Comment by Michiel29** [comment]
For anyone experiencing the same issue: my problem was that the index had a number of duplicate vectors, so one of the lists was very large. Removing these massively sped up the search. Closing
